### PR TITLE
multilabel for AveragePrecision

### LINKED ISSUE #359: Allow passing labels in (n_samples, n_classes) to AveragePrecision
## 🚀 Feature
Allow passing labels in `(n_samples, n_classes)` to [AveragePrecision](https://torchmetrics.readthedocs.io/en/latest/references/modules.html#averageprecision) like in sklearn.metrics [average_precision_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html)

### Motivation

```
ap = AveragePrecision(
    num_classes=7,
    pos_label=1,
    compute_on_step=False,
    dist_sync_on_step=False)
preds = torch.randn(8, 7)
target = torch.randint(0, 2, (8,7))
ap(preds, target)
ap.compute()
~/.env/lib/python3.8/site-packages/torchmetrics/functional/classification/precision_recall_curve.py in precision_recall_curve(preds, target, num_classes, pos_label, sample_weights)
    248         [tensor([0.7500]), tensor([0.7500]), tensor([0.0500, 0.7500]), tensor([0.0500, 0.7500]), tensor([0.0500])]
    249     """
--> 250     preds, target, num_classes, pos_label = _precision_recall_curve_update(preds, target, num_classes, pos_label)
    251     return _precision_recall_curve_compute(preds, target, num_classes, pos_label, sample_weights)

~/.env/lib/python3.8/site-packages/torchmetrics/functional/classification/precision_recall_curve.py in _precision_recall_curve_update(preds, target, num_classes, pos_label)
    107
    108     else:
--> 109         raise ValueError("preds and target must have same number of dimensions, or one additional dimension for preds")
    110
    111     return preds, target, num_classes, pos_label

ValueError: preds and target must have same number of dimensions, or one additional dimension for preds
```

### Pitch

In sklearn there is ability to calculate average precision for multilabel target:
```
from sklearn.metrics import average_precision_score
import numpy as np

target = np.random.randint(0, 2, (8, 7))
preds = np.random.randn(8, 7)

average_precision_score(target , preds)
0.505187074829932
```

### Additional context
Version:
```
torchmetrics==0.4.1
```

### PR DESCRIPTION
**Pull request by discort**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #359 

## PR review
@Borda @SkafteNicki 

## Did you have fun?
⚡⚡⚡

### CODE REVIEW

File: torchmetrics/functional/classification/precision_recall_curve.py [hunk 1]
```diff
@@ -148,13 +148,23 @@ def _precision_recall_curve_compute_multi_class(
...
+                sample_weights=sample_weights,
+            )
+        else:
+            res = precision_recall_curve(
+                preds=preds_cls,
+                target=target,
+                num_classes=1,
+                pos_label=cls,
+                sample_weights=sample_weights,
+            )
```
Comment (Borda) [inline torchmetrics/functional/classification/precision_recall_curve.py:167]:
```suggestion
        prc_args = dict(
                preds=preds_cls,
                target=target,
                num_classes=1,
                pos_label=cls,
                sample_weights=sample_weights,
            )
        if target.ndim > 1:
            prc_args.update(dict(precision_recall_curve(
                target=target[:, cls],
                pos_label=1,
            ))
        res = precision_recall_curve(**prc_args)
```

File: torchmetrics/functional/classification/precision_recall_curve.py [hunk 2]
```diff
@@ -148,13 +148,20 @@ def _precision_recall_curve_compute_multi_class(
...
+
+        prc_args = dict(
+                preds=preds_cls,
+                target=target,
+                num_classes=1,
+                pos_label=cls,
+                sample_weights=sample_weights,
+            )
+        if target.ndim > 1:
+            prc_args.update(dict(precision_recall_curve(
```
Comment (Borda) [inline torchmetrics/functional/classification/precision_recall_curve.py:160]:
```suggestion
            prc_args.update(dict(
```

**Review by Borda — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
nice!
it compiled without errors

File: tests/classification/test_average_precision.py [hunk 3]
```diff
@@ -118,3 +119,24 @@ def test_average_precision_differentiability(self, preds, sk_metric, target, num
 )
 def test_average_precision(scores, target, expected_score):
     assert average_precision(scores, target) == expected_score
+
+
+@pytest.mark.parametrize(
```
Comment (SkafteNicki) [inline tests/classification/test_average_precision.py:124]: I think it should be possible to just add an additional test case here:
https://github.com/PyTorchLightning/metrics/blob/80f654718bccf9d27094d6f8129d8a11b2041d9b/tests/classification/test_average_precision.py#L70
using this input case:
https://github.com/PyTorchLightning/metrics/blob/591bb2d7f92985efb321b54f96b2628c79e56fe5/tests/classification/inputs.py#L31-L33
(same as the one you already have)

**Review by SkafteNicki — CHANGES_REQUESTED** [review]
Implementation looks good, but could we try to fit this testcase inside the already established testing framework?

File: tests/classification/test_average_precision.py (hunk 3 above)
Comment (discort) [reply tests/classification/test_average_precision.py:124]: The initial idea was to test the same input  `(n_samples, n_classes)` such as for [sklearn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.average_precision_score.html).

`_input_binary_logits` has a bit differ shape `(NUM_BATCHES, BATCH_SIZE)`

Comment (SkafteNicki) [reply tests/classification/test_average_precision.py:124]: sorry, you are right, I did not account that the new test only checks a single batch. Then this should be the correct input:
https://github.com/PyTorchLightning/metrics/blob/591bb2d7f92985efb321b54f96b2628c79e56fe5/tests/classification/inputs.py#L50-L53

**Comment by Borda** [comment]
@SkafteNicki mind revisit?

**Review by SkafteNicki — APPROVED** [review]
LGTM :]

**Pull request merged**